Step 11.1 — Imports

In [1]:
import pandas as pd
import numpy as np
import joblib
import os

Step 11.2 — Load Final Model

In [2]:
final_model = joblib.load(
    "../models/final_fraud_model.pkl"
)

print("Final model loaded successfully.")
print(type(final_model))

Final model loaded successfully.
<class 'sklearn.linear_model._logistic.LogisticRegression'>


Step 11.3 — Load Final Threshold

In [3]:
threshold_config = pd.read_csv(
    "../models/final_model_threshold.csv"
)

threshold_config

,Model,Threshold
0,Logistic Regression,0.7


Step 11.4 — Extract Threshold

In [4]:
final_threshold = threshold_config.loc[
    0,
    "Threshold"
]

print("Final threshold:", final_threshold)

Final threshold: 0.7


Step 11.5 — Load Test Data

In [5]:
X_test = pd.read_csv(
    "../data/processed/X_test_processed.csv"
)

y_test = pd.read_csv(
    "../data/processed/y_test.csv"
).squeeze()

print("X_test shape:", X_test.shape)
print("y_test shape:", y_test.shape)

X_test shape: (1000, 30)
y_test shape: (1000,)


Step 11.6 — Verify Model Input Features

In [6]:
print("Number of model input features:", final_model.n_features_in_)
print("Number of test features:", X_test.shape[1])

Number of model input features: 30
Number of test features: 30


In [7]:

if final_model.n_features_in_ == X_test.shape[1]:
    print("Feature count verification passed.")
else:
    print("Feature count mismatch.")

Feature count verification passed.


Step 11.7 — Test Final Model

In [8]:
test_probabilities = final_model.predict_proba(
    X_test
)[:, 1]

test_predictions = (
    test_probabilities >= final_threshold
).astype(int)

print("Predictions generated successfully.")

Predictions generated successfully.


Step 11.8 — Verify Final Metrics

In [9]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

print("Final Model Performance")
print("=" * 40)

print(
    "Accuracy :",
    round(accuracy_score(y_test, test_predictions), 4)
)

print(
    "Precision:",
    round(
        precision_score(
            y_test,
            test_predictions,
            zero_division=0
        ),
        4
    )
)

print(
    "Recall   :",
    round(
        recall_score(
            y_test,
            test_predictions,
            zero_division=0
        ),
        4
    )
)

print(
    "F1 Score :",
    round(
        f1_score(
            y_test,
            test_predictions,
            zero_division=0
        ),
        4
    )
)

print(
    "ROC-AUC  :",
    round(
        roc_auc_score(
            y_test,
            test_probabilities
        ),
        4
    )
)

Final Model Performance
Accuracy : 0.885
Precision: 0.4317
Recall   : 0.625
F1 Score : 0.5106
ROC-AUC  : 0.7995


Step 11.9 — Save Model Metadata

In [10]:
model_metadata = {
    "model_name": "Logistic Regression",
    "model_file": "final_fraud_model.pkl",
    "classification_threshold": float(final_threshold),
    "number_of_features": int(final_model.n_features_in_),
    "accuracy": 0.8850,
    "precision": 0.431655,
    "recall": 0.6250,
    "f1_score": 0.510638,
    "roc_auc": 0.799514
}

model_metadata

{'model_name': 'Logistic Regression',
 'model_file': 'final_fraud_model.pkl',
 'classification_threshold': 0.7,
 'number_of_features': 30,
 'accuracy': 0.885,
 'precision': 0.431655,
 'recall': 0.625,
 'f1_score': 0.510638,
 'roc_auc': 0.799514}

Step 11.10 — Save Metadata

In [11]:
import json

with open(
    "../models/final_model_metadata.json",
    "w"
) as file:

    json.dump(
        model_metadata,
        file,
        indent=4
    )

print("Model metadata saved successfully.")

Model metadata saved successfully.


Step 11.11 — Final Verification

In [12]:
print("FINAL MODEL PACKAGE")
print("=" * 50)

print("Model:", model_metadata["model_name"])
print("Features:", model_metadata["number_of_features"])
print("Threshold:", model_metadata["classification_threshold"])

print("\nPerformance:")
print("Accuracy :", model_metadata["accuracy"])
print("Precision:", model_metadata["precision"])
print("Recall   :", model_metadata["recall"])
print("F1 Score :", model_metadata["f1_score"])
print("ROC-AUC  :", model_metadata["roc_auc"])

FINAL MODEL PACKAGE
Model: Logistic Regression
Features: 30
Threshold: 0.7

Performance:
Accuracy : 0.885
Precision: 0.431655
Recall   : 0.625
F1 Score : 0.510638
ROC-AUC  : 0.799514


Step 11.12 — Markdown

## Model Packaging Summary

The final selected fraud detection model was packaged together with its classification threshold and performance metadata.

The final model is Logistic Regression with a classification threshold of 0.70.

The model accepts 30 preprocessed input features and produces a fraud probability. A transaction is classified as fraudulent when its predicted fraud probability is greater than or equal to the selected threshold.

The packaged model will be used in the subsequent SQL, dashboard, transaction prediction, and fraud alert stages.